web3/utils/transactions.py

```diff
@@ -92,8 +92,19 @@ def get_required_transaction(web3, transaction_hash):
 
 
 def extract_valid_transaction_params(transaction_params):
-    return {key: transaction_params[key]
-            for key in VALID_TRANSACTION_PARAMS if key in transaction_params}
+    t_params = {key: transaction_params[key]
+                for key in VALID_TRANSACTION_PARAMS if key in transaction_params}
+
+    if t_params.get('data') is not None and transaction_params.get('input') is not None:
+        if t_params['data'] != transaction_params['input']:
+            msg = 'failure to handle this transaction due to both "input: {}" and'
+            msg += ' "data: {}" are populated. You need to resolve this conflict.'
+            raise AttributeError(msg.format(transaction_params['input'], t_params['data']))
```

Comment:
I find it easier to read when all of the alternative cases are explicitly handled in else blocks, even if it means a bit of duplication. eg~

```py
    if t_params.get('data') is not None and transaction_params.get('input') is not None:
        if t_params['data'] != transaction_params['input']:
            msg = 'failure to handle this transaction due to both "input: {}" and'
            msg += ' "data: {}" are populated. You need to resolve this conflict.'
            raise AttributeError(msg.format(transaction_params['input'], t_params['data']))
        else:
            return dissoc(t_params, 'input')
    elif t_params.get('data') is None and transaction_params.get('input') is not None:
        remapped_data =  t_params.pop('input')
        return assoc(
            dissoc(t_params, 'input'),
            'data',
            remapped_data,
        )
    else:
        return dissoc(t_params, 'input')
```

Reply:
Maybe even splitting up the `and` tests into nested ifs:
```py
if transaction_params.get('input') is None:
    return dissoc(t_params, 'input')
elif t_params.get('data') is None:
        ...
else:

Reply:
i didn't add *"input"* to *VALID_TRANSACTION_PARAMS*, so it's never in *extracted_params* (formerly *t_params*). instead, i'm checking  *transaction_params*  for *"input"* (throughout the function) so there's no managing of keys required; just value assignment or raise. do you want me to add it to the *VALID_TRANSACTION_PARAMS* ?